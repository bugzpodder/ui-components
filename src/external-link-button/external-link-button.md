###### Optional

- Any props for a Material UI Button will take, such as `className`

### Example

```js
const ExampleWrapper = require("@grail/components").ExampleWrapper;
const ExternalLinkButton = require("./external-link-button");

<ExampleWrapper>
  <ExternalLinkButton variant="raised" color="primary" href="https://www.grail.com">
    GRAIL Link in new tab
  </ExternalLinkButton>
</ExampleWrapper>;
```
