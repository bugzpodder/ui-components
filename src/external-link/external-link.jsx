// @flow
import React, { type Node } from "react";

type Props = {
	/** Takes a node to include in the omni dropdown after the search fields */
	children?: Node,
	/** The link */
	href: string,
};

/**
 *  ExternalLink opens in a new tab, while ensuring that the link protects from:
 * https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/
 */
export const ExternalLink = (props: Props) => {
	const { children, ...aTagProps } = props;
	return (
		<a
			{...aTagProps}
			target="_blank"
			rel="noopener noreferrer"
		>
			{children}
		</a>
	);
};
