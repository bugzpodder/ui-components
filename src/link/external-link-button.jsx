// @flow
import Button from "@material-ui/core/Button";
import React, { type ElementConfig, type Node } from "react";

type Props = {
  /** The children to display in the body of the button */
  children?: Node,
} & ElementConfig<typeof Button>;

/**
 *  ExternalLinkButton opens in a new tab, while ensuring that the link protects from:
 * https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/.
 * Similar to ExternalLink but it is represented in a button.
 */
export const ExternalLinkButton = (props: Props) => {
  const { children, ...buttonProps } = props;
  return (
    <Button
      data-testid="external-link-button"
      {...buttonProps}
      target="_blank"
      rel="noopener noreferrer"
    >
      {children}
    </Button>
  );
};
