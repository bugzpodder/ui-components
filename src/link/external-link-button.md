###### Optional

- Any props for a Material UI Button will take, such as `className`

### Example

```js
import { ExampleWrapper } from "../test-utils";
import { ExternalLinkButton } from "./";
<ExampleWrapper>
  <ExternalLinkButton variant="contained" color="primary" href="https://www.grail.com">
    GRAIL Link in new tab
  </ExternalLinkButton>
</ExampleWrapper>;
```
