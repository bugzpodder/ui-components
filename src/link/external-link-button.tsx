import Button, { ButtonProps } from "@material-ui/core/Button";

import React from "react";
import classNames from "classnames";
import styles from "./link-button.module.scss";

/**
 *  ExternalLinkButton opens in a new tab, while ensuring that the link protects from:
 * https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/.
 * Similar to ExternalLink but it is represented in a button.
 */
export const ExternalLinkButton: React.FC<ButtonProps> = props => {
  const { children, className, ...buttonProps } = props;
  const button = (
    // @ts-ignore property target does not exist.
    <Button
      data-testid="external-link-button"
      {...buttonProps}
      role="button"
      className={classNames(className, styles.link)}
      target="_blank"
      rel="noopener noreferrer"
    >
      {children}
    </Button>
  );
  return <div className={styles.button}>{button}</div>;
};
