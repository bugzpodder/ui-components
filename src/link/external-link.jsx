// @flow
import React, { type Node } from "react";

type Props = {
  /** Takes a node to include in the omni dropdown after the search fields */
  children?: Node,
  /** The link */
  href: string,
  /** data-testid for the component */
  "data-testid"?: string,
};

/**
 *  ExternalLink opens in a new tab, while ensuring that the link protects from:
 * https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/
 */
export const ExternalLink = React.forwardRef<Props, any>((props: Props, ref: any) => {
  const { children, "data-testid": dataTestId = "external-link", ...aTagProps } = props;
  return (
    <a
      ref={ref}
      data-testid={dataTestId}
      {...aTagProps}
      target="_blank"
      rel="noopener noreferrer"
    >
      {children}
    </a>
  );
});
