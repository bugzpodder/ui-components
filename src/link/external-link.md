###### Optional

- Any props that an `a` tag will take, such as `className`

### Example

```js
<ExternalLink href="https://www.grail.com">GRAIL Link in new tab</ExternalLink>
```
