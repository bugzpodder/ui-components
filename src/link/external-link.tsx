import React, { forwardRef } from "react";

type Props = {
  /** data-testid for the component */
  "data-testid"?: string;
  /** className for the component */
} & React.AnchorHTMLAttributes<HTMLAnchorElement>;

/**
 *  ExternalLink opens in a new tab, while ensuring that the link protects from:
 * https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/
 */
export const ExternalLink = forwardRef<any, Props>((props: Props, ref: any) => {
  const {
    children,
    "data-testid": dataTestId = "external-link",
    ...aTagProps
  } = props;
  return (
    <a
      ref={ref}
      data-testid={dataTestId}
      {...aTagProps}
      target="_blank"
      rel="noopener noreferrer"
    >
      {children}
    </a>
  );
});
