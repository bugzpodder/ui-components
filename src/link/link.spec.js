// @flow
import "jest-dom/extend-expect";
import React from "react";
import { ExternalLink } from "./external-link";
import { ExternalLinkButton } from "./external-link-button";
import { TestWrapper } from "../test-utils";
import { cleanup, render } from "@testing-library/react";

afterEach(cleanup);

test("render ExternalLink", () => {
  const { container, getByTestId } = render(
    <TestWrapper>
      <ExternalLink
        href="https://www.grail.com"
        className="test"
      >
        GRAIL
      </ExternalLink>
    </TestWrapper>,
  );
  expect(container).toMatchSnapshot();
  expect(getByTestId("external-link")).toBeInTheDocument();
  expect(getByTestId("external-link")).toHaveAttribute("href", "https://www.grail.com");
  expect(getByTestId("external-link")).toHaveAttribute("rel", "noopener noreferrer");
  expect(getByTestId("external-link")).toHaveAttribute("target", "_blank");
  expect(getByTestId("external-link")).toHaveTextContent("GRAIL");
});

test("render ExternalLink", () => {
  const { container, getByTestId } = render(
    <TestWrapper>
      <ExternalLinkButton
        href="https://www.grail.com"
        className="test"
      >
        GRAIL
      </ExternalLinkButton>
    </TestWrapper>,
  );
  expect(container).toMatchSnapshot();
  expect(container).toMatchSnapshot();
  expect(getByTestId("external-link-button")).toBeInTheDocument();
  expect(getByTestId("external-link-button")).toHaveAttribute("href", "https://www.grail.com");
  expect(getByTestId("external-link-button")).toHaveAttribute("rel", "noopener noreferrer");
  expect(getByTestId("external-link-button")).toHaveAttribute("target", "_blank");
  expect(getByTestId("external-link-button")).toHaveTextContent("GRAIL");
});
