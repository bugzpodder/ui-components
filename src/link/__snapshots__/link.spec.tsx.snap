// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExternalLink 1`] = `
<div>
  <a
    data-testid="external-link"
    href="https://www.grail.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    GRAIL
  </a>
</div>
`;

exports[`render ExternalLink 2`] = `
<div>
  <div
    class="button"
  >
    <a
      aria-disabled="false"
      class="test link"
      data-testid="external-link-button"
      href="https://www.grail.com"
      rel="noopener noreferrer"
      role="button"
      tabindex="0"
      target="_blank"
    >
      <span>
        GRAIL
      </span>
      <span
        class=""
      />
    </a>
  </div>
</div>
`;

exports[`render ExternalLink 3`] = `
<div>
  <div
    class="button"
  >
    <a
      aria-disabled="false"
      class="test link"
      data-testid="external-link-button"
      href="https://www.grail.com"
      rel="noopener noreferrer"
      role="button"
      tabindex="0"
      target="_blank"
    >
      <span>
        GRAIL
      </span>
      <span
        class=""
      />
    </a>
  </div>
</div>
`;
